### PR TITLE
remove path from failpoint

### PR DESCRIPTION
**Pull request by lavacat**
See https://github.com/etcd-io/etcd/issues/14727

### CODE REVIEW

**Comment by serathius** [comment]
This looks good for me. cc @ahrtr who had different opinion about the topic. 
Can you fix the DCO check? Just run `git checkout dd4869804163318da0745652c38501072f3c824d && git commit --amend --signoff && git push HEAD:remove-fp-path -f`

**Comment by ahrtr** [comment]
> cc @ahrtr who had different opinion about the topic.

Just as I mentioned in https://github.com/etcd-io/etcd/issues/14727#issuecomment-1329654978, it isn't a big deal. Since no feedback from other maintainers. I don't want to be blocked here. I agree to remove the pacakge path, although personally I tend to keep it.

File: runtime/runtime.go [hunk 1]
```diff
@@ -147,11 +147,17 @@ func list() []string {
 	return ret
 }
 
-func register(failpath string, fp *Failpoint) {
+func register(name string, fp *Failpoint) {
 	failpointsMu.Lock()
```
[the remark this answers is not in the document]
Comment (ahrtr) [reply runtime/runtime.go:151]: Note that we remove the lock for each failpoint. We need to acquire the global lock for each operation.

File: runtime/runtime.go [hunk 2]
```diff
@@ -147,11 +147,17 @@ func list() []string {
...
-	failpoints[failpath] = fp
+	if _, ok := failpoints[name]; ok {
+		fmt.Fprintf(
+			os.Stderr,
+			"warning: failpoint name %s is already registered. You are using same failpoint name in multiple packages.\n",
+			name)
+	}
+	failpoints[name] = fp
 	failpointsMu.Unlock()
```
Comment (ahrtr) [inline runtime/runtime.go:159]:
```suggestion
```

**Comment by ahrtr** [comment]
We don't have e2e test for now, please make sure you manually verify the PR using [examples](https://github.com/etcd-io/gofail/tree/master/examples)

**Comment by serathius** [comment]
@lavacat Can you apply the suggestions by @ahrtr ?

**Comment by lavacat** [comment]
- added unit tests 
- renamed `failpath` to `name` everywhere in runtime.go

@ahrtr FYI with your suggestion there was a deadlock, because register was calling `Enable` that acquires the same lock. Changed it to call `enable`. I think that was a typo in original code.

**Comment by lavacat** [comment]
- changed NewFailpoint/register logic to return existing failpoint if name is already registered
- updated readme
- minor unit tests improvements
@serathius @ahrtr I think this is ready to merge. Please take a look.

File: runtime/runtime.go [hunk 3]
```diff
@@ -147,11 +147,22 @@ func list() []string {
...
-	if t, ok := envTerms[failpath]; ok {
-		Enable(failpath, t)
+	defer failpointsMu.Unlock()
+	if f, ok := failpoints[name]; ok {
+		fmt.Fprintf(
+			os.Stderr,
+			"warning: failpoint name %s is already registered. You are using same failpoint name in multiple packages.\n",
+			name)
+		// return existing failpoint, don't re-enable
+		return f
```
Comment (ahrtr) [inline runtime/runtime.go:159]: We shouldn't allow duplicated failpoint, so we might need to panic here? WDYT? cc @serathius

**Comment by ahrtr** [comment]
Overall looks good to me with two comments:
1. Please fix the test failure;
2. Please consider using `github.com/stretchr/testify` in the unit test so simplify the implementation.

File: runtime/runtime.go (hunk 3 above)
Comment (serathius) [reply runtime/runtime.go:159]: Yep, we definitely need to panic here.

**Comment by lavacat** [comment]
> Overall looks good to me with two comments:
> 
> 1. Please fix the test failure;
> 2. Please consider using `github.com/stretchr/testify` in the unit test so simplify the implementation.

@ahrtr thanks, addressed both

File: code/binding_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package code
+
+import (
+	"bytes"
+	"github.com/stretchr/testify/assert"
+	"testing"
```
Comment (ahrtr) [inline code/binding_test.go:20]:
```suggestion
	"bytes"
	"testing"
	
	"github.com/stretchr/testify/assert"
```

File: runtime/failpoint_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
```
Comment (ahrtr) [inline runtime/failpoint_test.go:19]:
```suggestion
	"testing"
	
	"github.com/stretchr/testify/assert"
```

File: runtime/runtime.go [hunk 6]
```diff
@@ -147,11 +147,19 @@ func list() []string {
...
+func register(name string) *Failpoint {
 	failpointsMu.Lock()
-	failpoints[failpath] = fp
-	failpointsMu.Unlock()
-	if t, ok := envTerms[failpath]; ok {
-		Enable(failpath, t)
+	defer failpointsMu.Unlock()
+	if _, ok := failpoints[name]; ok {
+		panic(fmt.Sprintf(
+			"failpoint name %s is already registered. You are using same failpoint name in multiple packages.\n",
```
Comment (serathius) [inline runtime/runtime.go:155]:
```suggestion
			"failpoint name %s is already registered.",
```
Could be same package.

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you @lavacat

**Pull request merged**